connector/saml/saml.go

```diff
@@ -348,19 +355,151 @@ func (p *provider) HandlePOST(s connector.Scopes, samlResponse string) (ident co
 	return ident, nil
 }
 
+// Validate that the StatusCode of the Response is success.
+// Otherwise return a human readable message to the end user
+func (p *provider) validateStatus(resp *response) error {
+	// Status is mandatory in the Response type
+	status := resp.Status
+	if status == nil {
+		return fmt.Errorf("response did not contain a Status")
+	}
+	// StatusCode is mandatory in the Status type
+	statusCode := status.StatusCode
+	if statusCode == nil {
+		return fmt.Errorf("response did not contain a StatusCode")
+	}
+	if statusCode.Value != statusCodeSuccess {
+		parts := strings.Split(statusCode.Value, ":")
+		lastPart := parts[len(parts)-1]
+		errorMessage := fmt.Sprintf("The status code of the Response was not Success, was %q", lastPart)
+		statusMessage := status.StatusMessage
+		if statusMessage != nil && statusMessage.Value != "" {
+			errorMessage += " -> " + statusMessage.Value
+		}
+		return fmt.Errorf(errorMessage)
+	}
+	return nil
+}
+
+// Multiple subject SubjectConfirmation can be in the assertion
+// and at least one SubjectConfirmation must be valid.
+// This is described in the spec "Profiles for the OASIS Security
+// Assertion Markup Language" in section 3.3 Bearer.
+// see https://www.oasis-open.org/committees/download.php/35389/sstc-saml-profiles-errata-2.0-wd-06-diff.pdf
+func (p *provider) validateSubjectConfirmation(subject *subject) error {
+	validSubjectConfirmation := false
+	subjectConfirmations := subject.SubjectConfirmations
+	if subjectConfirmations != nil && len(subjectConfirmations) > 0 {
+		for _, subjectConfirmation := range subjectConfirmations {
+			// skip if method is wrong
+			method := subjectConfirmation.Method
+			if method != "" && method != subjectConfirmationMethodBearer {
+				continue
+			}
+			subjectConfirmationData := subjectConfirmation.SubjectConfirmationData
+			if subjectConfirmationData == nil {
+				continue
+			}
+			inResponseTo := subjectConfirmationData.InResponseTo
+			if inResponseTo != "" {
+				// TODO also validate InResponseTo if present
+			}
+			// only validate that subjectConfirmationData is not expired
+			now := p.now()
+			notOnOrAfter := time.Time(subjectConfirmationData.NotOnOrAfter)
+			if !notOnOrAfter.IsZero() {
+				if now.After(notOnOrAfter) {
```

Comment:
They don't do it in the [python-saml](https://github.com/onelogin/python-saml/blob/master/src/onelogin/saml2/response.py#L252-L256). I don't now why but wanted to be in line? For the expiration case `NotOnOrAfter` the problem is anyway not so critical as for the not valid yet case which I currently do not validate for subjectConfirmations.